block/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -72,16 +83,67 @@ retrievalWorker keepAliveTimer = worker outs (retrievalWorkerImpl keepAliveTimer
 --
 -- If both happen at the same time, 'BlockRetrievalQueue' takes precedence.
 --
+-- Note: if we haven't caught up with the blockchain yet, the worker will
+-- try to download and apply blockchain dumps, and only if it fails it'll
+-- switch to requesting headers/blocks.
+--
 retrievalWorkerImpl
     :: forall ctx m.
        (BlockWorkMode ctx m)
-    => Timer -> SendActions m -> m ()
-retrievalWorkerImpl keepAliveTimer SendActions {..} =
-    handleAll mainLoopE $ do
-        logInfo "Starting retrievalWorker loop"
-        mainLoop
+    => Timer -> Maybe Text -> SendActions m -> m ()
+retrievalWorkerImpl keepAliveTimer blockStorageMirror SendActions {..} = do
+    handleAll tryDownloadE $ do
+        epoch <- view epochIndexL <$> DB.getTipHeader
+        whenJust blockStorageMirror $ \dumpUrl ->
+            tryDownload epoch (toString dumpUrl)
+    handleAll recoveryLoopE $ do
+        logInfo "Starting recovery loop"
+        recoveryLoop
   where
-    mainLoop = do
+    -- | Try to download and apply a blockchain dump, starting from the
+    -- given epoch.
+    tryDownload epoch dumpUrl = whenNothingM_ getCurrentSlot $ do
```

Comment:
Please, move it to the separate function. `where` section of the worker is too big already.

Reply:
@Pastafarianist @volhovm fight

Reply:
There's no need to fight, you're moving it to the separate function :peka:

Reply:
I want you to fight because @Pastafarianist believes that stuff should be in `where` and I want him to stop believing it. (As for me, I'm obviously going to move it to a separate function.)

Reply:
Ah, the scope locality debate. Why don't we just rewrite the whole thing in C++ so that we could use *namespaces*?

Reply:
> Ah, the scope locality debate.

To be honest, I don't see a particularly good reason to spend effort / hurt readability in order to preserve the “no more things in scope than absolutely necessary” thing. Sure it feels *nice* (just like it feels nice to not use `WorkMode` when a smaller set of constraints suffices), but that's not a particularly good reason.